File: mycodes/42-TrappingRainWater.java
class Solution {
    public int trap(int[] height) {
        if (height == null || height.length == 0) return 0;
        int max = height[0], maxIndex = 0;
        for (int i = 1; i < height.length; ++i) {
            if (max < height[i]) {
                max = height[i];
                maxIndex = i;
            }
        }
        int maximal = 0, res = 0;
        for (int i = 0; i < maxIndex; ++i) {
            if (height[i] < maximal) res += maximal - height[i];
            else maximal = height[i];
        }
        maximal = 0;
        for (int i = height.length - 1; i > maxIndex; --i) {
            if (height[i] < maximal) res += maximal - height[i];
            else maximal = height[i];
        }
        return res;
    }
}

File: mycodes/42-TrappingRainWater.go
package main

func trap(height []int) int {
  wall := make([]int, 0, len(height))
  res := 0
  maxHeight := 0
  for i := range height {
    h := min(maxHeight, height[i])
    for j := len(wall)-1; j >= 0; j-- {
      if wall[j] < h {
        res += h - wall[j]
        wall[j] = h
      } else {
        break
      }
    }
    wall = append(wall, height[i])
    maxHeight = max(maxHeight, height[i])
  }
  return res
}
